### PR TITLE
Fix suggest features for narrow stacks

### LINKED ISSUE #1280: Automated feature selection should gracefully handle data sets that are too small for the largest filters in at least one dimension
> When I first ran Ilastik, I loaded my images and selected all features.
> Then, Ilastik complained that the largest (sigma 10) features are too 
> big for my data. No problem with that, just deselected it.
> 
> But when I want to run the "suggest features", I get the very same 
> error message.

### PR DESCRIPTION
**Pull request by k-dominik**
Enable _SuggestFeatures_ Dialog to deal with "narrow" stacks. Previously it would error out and might even produce segfaults :o.

On the way there was
* refactoring - naming _is_ important. It is frustrating every time I am looking for this particular dialog. Now naming reflects what is shown in the UI.
* made the dialog properly modal and made it run in the same eventloop
* made its `_run_selection` method somewhat more robust to errors - at least the graph is reverted to the same state as before
* by adding some attributes to the error that is raised by the `OpFeatureSelection` when incompatible scales are used, the dialog now automatically deselects the ones that do not work (for features that only don't fit in _z_, they are now calculated in 2D).

See also #2637 

Bottom line: Ideally this whole functionality would be reworked from the ground up. This is just a patch.

fixes #1280 

## Checklist

- [x] Format code and imports.
- [x] Add docstrings and comments.
- [x] Add tests.
- [x] Reference relevant issues and other pull requests.
- [x] Rebase commits into a logical sequence.

### CODE REVIEW

File: tests/test_ilastik/test_applets/pixelClassification/testPixelClassificationGui.py [hunk 1]
```diff
@@ -266,8 +267,8 @@ def impl():
             selections = numpy.array(
                 [
                     [True, False, False, False, False, False, False],
-                    [True, False, False, False, False, False, False],
-                    [True, False, False, False, False, False, False],
+                    [False, True, False, False, False, False, False],
+                    [False, True, False, False, False, False, False],
```
Comment (k-dominik) [inline tests/test_ilastik/test_applets/pixelClassification/testPixelClassificationGui.py:271]: previous feature set was "illegal" for ilastik. It is not possible to achieve using the gui.

But main reason for changing it is that the SuggestFeatures Dialog that later is used in this test, makes this assumption

File: ilastik/applets/pixelClassification/suggestFeaturesDialog.py [hunk 2]
```diff
@@ -767,157 +777,186 @@ def _auto_select_num_features(self, feature_order):
...
+                    compute_in_2d[numpy.isin(scales, err.invalid_z_scales)] = True
+                    self.opFeatureSelection.ComputeIn2d.setValue(compute_in_2d.tolist())
+                    self.compute_in_2d_compat = compute_in_2d.tolist()
+                    msg += f"Calculating features at sigmas {err.invalid_z_scales} in 2D only."
+
+                logger.info(msg)
+
+                # suppress the popup in pytest:
+                if "PYTEST_CURRENT_TEST" not in os.environ:
+                    QMessageBox.information(self, "Not all features are compatible with your data!", msg)
```
Comment (k-dominik) [inline ilastik/applets/pixelClassification/suggestFeaturesDialog.py:819]: I am unhappy to have pytest-related shenanigans in production code, but alternative would be to keep a handle to the message box around in the dialog itself (I didn't manage to get a handle on it with `.activeModalWidget()`...

File: tests/test_ilastik/test_applets/pixelClassification/testPixelClassificationGui.py [hunk 3]
```diff
@@ -678,3 +679,76 @@ def impl():
...
+            gui = pixClassApplet.getMultiLaneGui()
+            opFeatures = workflow.featureSelectionApplet.topLevelOperator
+
+            # this time we want to trigger the message box that informs us,
+            # that some sigmas can only be computed in 2D
+            # so we add a sigma that doesn't fit into size of 50 along z:
+            scales = [0.3, 0.7, 15.0]
+            selections = numpy.array(
+                [
+                    [True, False, False],
```
Comment (emilmelnikov) [inline tests/test_ilastik/test_applets/pixelClassification/testPixelClassificationGui.py:718]: A readability idea: use `0` and `1` instead of `False` and `True`, and then `.astype(bool)` the array. Single-digit numbers are shorter and line up better than boolean constants. Could use this trick in other places too.

File: tests/test_ilastik/test_applets/pixelClassification/testPixelClassificationGui.py [hunk 4]
```diff
@@ -678,3 +679,76 @@ def impl():
...
+                    [False, True, False],
+                    [False, False, False],
+                    [False, False, False],
+                    [False, False, False],
+                ]
+            )
+
+            opFeatures.SelectionMatrix.setValue(None)
+            opFeatures.Scales.setValue(scales)
+            opFeatures.ComputeIn2d.setValue([False for _ in scales])
```
Comment (emilmelnikov) [inline tests/test_ilastik/test_applets/pixelClassification/testPixelClassificationGui.py:729]:
```suggestion
            opFeatures.ComputeIn2d.setValue([False] * len(scales))
```

File: ilastik/applets/featureSelection/__init__.py [hunk 5]
```diff
@@ -20,4 +18,44 @@
...
+_SCALES = Iterable[Union[int, float]]
+
+
+class FeatureSelectionConstraintError(DatasetConstraintError):
+    def __init__(
+        self,
+        appletName: str,
+        invalid_scales: _SCALES,
+        invalid_z_scales: _SCALES,
+        fixing_dialogs=Iterable[Callable],
```
Comment (emilmelnikov) [inline ilastik/applets/featureSelection/__init__.py:34]: Default arguments for a seemingly common case:
```suggestion
        invalid_z_scales: _SCALES = (),
        fixing_dialogs=Iterable[Callable] = (),
```

File: ilastik/applets/pixelClassification/suggestFeaturesDialog.py [hunk 6]
```diff
@@ -1,33 +1,33 @@
...
-import re
+
+import logging
+
+logger = logging.getLogger(__name__)
 
 
 # just a container class, nothing fancy here
-class FeatureSelectionResult(object):
+class SuggestFeaturesResult(object):
```
Comment (emilmelnikov) [inline ilastik/applets/pixelClassification/suggestFeaturesDialog.py:30]:
```suggestion
class SuggestFeaturesResult:
```

File: ilastik/applets/pixelClassification/suggestFeaturesDialog.py [hunk 7]
```diff
@@ -767,157 +777,186 @@ def _auto_select_num_features(self, feature_order):
...
+                )
+                self._add_feature_set_to_results(current_features_result)
+                self._initialized_current_features_segmentation_layer = True
 
-        # revert changes to matrix
-        self.opFeatureSelection.SelectionMatrix.setValue(user_defined_matrix)
-        self.opFeatureSelection.SelectionMatrix.setDirty()  # this does not do anything!?!?
-        self.opFeatureSelection.setupOutputs()
-        QtWidgets.QApplication.instance().restoreOverrideCursor()
+        finally:
```
Comment (emilmelnikov) [inline ilastik/applets/pixelClassification/suggestFeaturesDialog.py:941]: I wonder if a context manager that reverts slot values on `__exit__` would be more readable and robust.

**Review by emilmelnikov — COMMENTED** [review]
I've reviewed some parts of the new code, but left alone other parts that were unchanged, because we agree that it required a rewrite anyway.

File: ilastik/applets/featureSelection/__init__.py [hunk 8]
```diff
@@ -20,4 +18,44 @@
 # on the ilastik web site at:
 # 		   http://ilastik.org/license.html
 ###############################################################################
+from typing import Callable, Iterable, Union
+
+from ilastik.applets.base.applet import DatasetConstraintError
+
+_SCALES = Iterable[Union[int, float]]
```
[the remark this answers is not in the document]
Comment (k-dominik) [reply ilastik/applets/featureSelection/__init__.py:25]: thx!

File: ilastik/applets/pixelClassification/suggestFeaturesDialog.py (hunk 7 above)
Comment (k-dominik) [reply ilastik/applets/pixelClassification/suggestFeaturesDialog.py:941]: agreed, I'll give it a go and see how it looks. Thx! :)

**Comment by k-dominik** [comment]
thx for the review @emilmelnikov
